hs_core/views/resource_rest_api.py

```diff
@@ -0,0 +1,622 @@
+from rest_framework.views import APIView
+from rest_framework.response import Response
+from rest_framework import status, generics
+from rest_framework.exceptions import *
+from hs_core import hydroshare
+from hs_core.hydroshare import file_size_limit_for_display
+from hs_core.views import utils
+import simplejson as sj
+import serializers
+from hs_core.signals import *
+import os
+
+class ResourceRetrieveUpdateDelete(APIView):
+    """
+    Retrieve, update or delete a resource
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: GET
+
+    HTTP status code (on success):200
+
+    :return: (on success): resource bag file if request content-type is 'application/zip' otherwise resource science metadata file
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: DELETE
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    REST URL: hsapi/resource/{pk}
+
+    HTTP method: PUT
+
+    HTTP status code (on success):200
+
+    :return: (on success): json string of the format: {'resource_id':pk}
+
+    :type   str
+
+    :param  pk: resource id
+
+    :rtype:  json string for http methods DELETE and PUT, and resource file data bytes for GET
+
+    :raises:
+
+
+    &nbsp;&nbsp;&nbsp;NotFound: return json format: {'detail': 'No resource was found for resource id':pk}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code: 404
+
+    &nbsp;&nbsp;&nbsp;PermissionDenied: return json format: {'detail': 'You do not have permission to perform this action.'}
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:403
+
+    &nbsp;&nbsp;&nbsp;ValidationError: return json format: {parameter-1': ['error message-1'], 'parameter-2': ['error message-2'], .. }
+
+    &nbsp;&nbsp;&nbsp;HTTP status code:400
+    """
+
+    @property
+    def allowed_methods(self):
+        return ['GET', 'PUT', 'DELETE']
+
+    def get(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, view=True)
+        resource_bag = hydroshare.get_resource(pk)
+
+        if request.content_type == 'application/zip':
+            # redirects to django_irods/views.download function
+            return HttpResponseRedirect(resource_bag.bag.url)
+        else:
+            return redirect('science_metadata_get_update', pk=pk)
+
+    def put(self, request, pk):
+        # check the user is authorized for this action and if the resource exists for the given resource id (pk)
+        utils.authorize(request, pk, edit=True)
+
+        resource_update_request_validator = serializers.ResourceUpdateRequestValidator(data=request.data)
+        if not resource_update_request_validator.is_valid():
+            raise ValidationError(detail=resource_update_request_validator.errors)
+
+        # TODO: update resource - not sure what exactly we need to update
+        raise NotImplementedError()
```

Comment:
@horsburgh You raise a good point.  I'm not sure there is one right answer though.  Some resources may refer to all files in a resource when creating/updating metadata.  Other resources may only need to look at the first file to determine the resource's science metadata, and yet others may not care about the files at all.

One way to handle this would be to rely on the logic in pre_create_resource signal handler that resources can define.  We ought to be able to send this signal to the resource.  The signal handler will give us the appropriate metadata (in the dictionary reference that we pass via kwargs['metadata']), which we can use to update the metadata.  For this to work as expected (i.e. to behave the same way as during resource creation), we will have to ensure that the file being updated is passed to pre_create_resource in the same order as it was passed when the resource was created.  That way, if only the first N files are used to determine metadata, the metadata will be properly updated if the file being overwritten was or was not the Nth file. 

If we go this route, we will need to verify that all current resources are only using pre_create_resource signal handlers to determine metadata, and that these signal handlers have no other side effects.  If there are other things we might want to allow a resource to do in pre_create_resource handlers, we may want to define a second signal that only handles metadata extraction, and ensure that such signals are raised from pre_create_resource.

Does this make sense, or am I missing something important?

@hyi What is your opinion of this solution?